### PR TITLE
New Resource: azurerm_security_center_assessment_metadata

### PR DESCRIPTION
**Pull request by neil-yechenwei**
Currently, terraform doesn't support Security.AssessmentMetadata RP. So submitted this PR to implement it.

--- PASS: TestAccSecurityCenterAssessmentMetadata_basic (66.90s)
--- PASS: TestAccSecurityCenterAssessmentMetadata_complete (68.07s)
--- PASS: TestAccSecurityCenterAssessmentMetadata_update (125.80s)

The api in go sdk package:
https://github.com/Azure/azure-sdk-for-go/blob/master/services/preview/security/mgmt/v3.0/security/assessmentsmetadata.go

### CODE REVIEW

**Review by katbyte — CHANGES_REQUESTED** [review]
We have some test failing:
```
------- Stdout: -------
=== RUN   TestAccSecurityCenterAssessmentMetadata_complete
=== PAUSE TestAccSecurityCenterAssessmentMetadata_complete
=== CONT  TestAccSecurityCenterAssessmentMetadata_complete
    testing.go:745: Error destroying resource! WARNING: Dangling resources
        may exist. The full state and error is shown below.
        
        Error: Check failed: "azurerm_security_center_assessment_metadata.test" still exists
        
        State: <no state>
--- FAIL: TestAccSecurityCenterAssessmentMetadata_complete (43.00s)
FAIL

------- Stdout: -------
=== RUN   TestAccSecurityCenterAssessmentMetadata_update
=== PAUSE TestAccSecurityCenterAssessmentMetadata_update
=== CONT  TestAccSecurityCenterAssessmentMetadata_update
    testing.go:684: Step 0 error: Check failed: Check 1/1 error: "azurerm_security_center_assessment_metadata.test" did not exist
--- FAIL: TestAccSecurityCenterAssessmentMetadata_update (22.06s)
FAIL
```

**Comment by neil-yechenwei** [comment]
@katbyte , After rerun, seems all test cases passed on my local. Maybe rerun can resolve your problem?

$ make testacc TEST=./azurerm/internal/services/securitycenter/ TESTARGS="-parallel 11 -test.run=TestAccSecurityCenterAssessmentMetadata_"
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test ./azurerm/internal/services/securitycenter/ -v -parallel 11 -test.run=TestAccSecurityCenterAssessmentMetadata_ -timeout 180m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccSecurityCenterAssessmentMetadata_basic
=== PAUSE TestAccSecurityCenterAssessmentMetadata_basic
=== RUN   TestAccSecurityCenterAssessmentMetadata_requiresImport
=== PAUSE TestAccSecurityCenterAssessmentMetadata_requiresImport
=== RUN   TestAccSecurityCenterAssessmentMetadata_complete
=== PAUSE TestAccSecurityCenterAssessmentMetadata_complete
=== RUN   TestAccSecurityCenterAssessmentMetadata_update
=== PAUSE TestAccSecurityCenterAssessmentMetadata_update
=== CONT  TestAccSecurityCenterAssessmentMetadata_basic
=== CONT  TestAccSecurityCenterAssessmentMetadata_update
=== CONT  TestAccSecurityCenterAssessmentMetadata_complete
=== CONT  TestAccSecurityCenterAssessmentMetadata_requiresImport
--- PASS: TestAccSecurityCenterAssessmentMetadata_basic (64.15s)
--- PASS: TestAccSecurityCenterAssessmentMetadata_complete (65.74s)
--- PASS: TestAccSecurityCenterAssessmentMetadata_requiresImport (79.13s)
--- PASS: TestAccSecurityCenterAssessmentMetadata_update (116.31s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/securitycenter      121.073s

**Comment by neil-yechenwei** [comment]
@katbyte , may I ask whether the test cases are passed from your side?

**Comment by neil-yechenwei** [comment]
@katbyte , Seems test case can pass after rerun.

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hi @neil-yechenwei 

Thanks for this PR - I've taken a look through and left some comments inline - since several of these fields can't be (or don't make sense to be, in the case of the uuid) user specified, it'd be worth rethinking the design here (since we can generate/default values for these as necessary) - but if we can fix those up then we should be able to take another look.

Thanks!

File: azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+
+	if d.IsNewResource() {
+		existing, err := client.GetInSubscription(ctx, name)
+		if err != nil {
+			if !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("checking for present of existing Security Center Assessment Metadata %q : %+v", name, err)
+			}
+		}
+
+		if existing.ID != nil && *existing.ID != "" {
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go:152]: updated

File: azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+			DisplayName:    utils.String(d.Get("display_name").(string)),
+			Severity:       security.Severity(d.Get("severity").(string)),
+		},
+	}
+
+	if v, ok := d.GetOk("threats"); ok {
+		threats := make([]security.Threats, 0)
+		for _, item := range *(utils.ExpandStringSlice(v.(*schema.Set).List())) {
+			threats = append(threats, (security.Threats)(item))
+		}
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go:170]: updated

File: azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+		d.Set("severity", props.Severity)
+		d.Set("implementation_effort", props.ImplementationEffort)
+		d.Set("is_preview", props.Preview)
+		d.Set("remediation_description", props.RemediationDescription)
+		d.Set("user_impact", props.UserImpact)
+
+		threats := make([]string, 0)
+		if props.Threats != nil {
+			for _, item := range *props.Threats {
+				threats = append(threats, (string)(item))
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go:234]: updated

File: azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+	ctx, cancel := timeouts.ForDelete(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := parse.AssessmentMetadataID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	if _, err := client.DeleteInSubscription(ctx, id.AssessmentMetadataName); err != nil {
+		return fmt.Errorf("deleting Security Center Assessment Metadata %q : %+v", id.AssessmentMetadataName, err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go:254]: updated

File: azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+	}
+
+	resp, err := client.GetInSubscription(ctx, id.AssessmentMetadataName)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[INFO] security Center Assessment Metadata %q does not exist - removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("retrieving Security Center Assessment Metadata %q : %+v", id.AssessmentMetadataName, err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go:216]: updated

File: azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+
+	id, err := parse.AssessmentMetadataID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	resp, err := client.GetInSubscription(ctx, id.AssessmentMetadataName)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[INFO] security Center Assessment Metadata %q does not exist - removing from state", d.Id())
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go:212]: updated

File: website/docs/r/security_center_assessment_metadata.html.markdown [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+The `timeouts` block allows you to specify [timeouts](https://www.terraform.io/docs/configuration/resources.html#timeouts) for certain actions:
+
+* `create` - (Defaults to 30 minutes) Used when creating the Security Center Assessment Metadata.
+* `read` - (Defaults to 5 minutes) Used when retrieving the Security Center Assessment Metadata.
+* `update` - (Defaults to 30 minutes) Used when updating the Security Center Assessment Metadata.
+* `delete` - (Defaults to 30 minutes) Used when deleting the Security Center Assessment Metadata.
+
+## Import
+
+security AssessmentsMetadatums can be imported using the `resource id`, e.g.
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply website/docs/r/security_center_assessment_metadata.html.markdown:68]: updated

File: website/docs/r/security_center_assessment_metadata.html.markdown [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+* `name` - (Required) The GUID as name which should be used for this Security Center Assessment Metadata. Changing this forces a new Security Center Assessment Metadata to be created.
+
+* `assessment_type` - (Required) The type of the Security Center Assessment which can be specified by user. Possible value is `CustomerManaged`.
+
+* `description` - (Required) The description of the Security Center Assessment.
+
+* `display_name` - (Required) The user-friendly display name of the Security Center Assessment.
+
+* `severity` - (Required) The severity level of the Security Center Assessment. Possible values are `Low`, `Medium` and `High`.
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply website/docs/r/security_center_assessment_metadata.html.markdown:37]: updated

File: website/docs/r/security_center_assessment_metadata.html.markdown [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+}
+```
+
+## Arguments Reference
+
+The following arguments are supported:
+
+* `name` - (Required) The GUID as name which should be used for this Security Center Assessment Metadata. Changing this forces a new Security Center Assessment Metadata to be created.
+
+* `assessment_type` - (Required) The type of the Security Center Assessment which can be specified by user. Possible value is `CustomerManaged`.
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply website/docs/r/security_center_assessment_metadata.html.markdown:31]: I am not sure whether more assessment types would be introduced. So I expose it here. Does it make sense?

File: azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+
+	if props := resp.AssessmentMetadataProperties; props != nil {
+		d.Set("assessment_type", props.AssessmentType)
+		d.Set("description", props.Description)
+		d.Set("display_name", props.DisplayName)
+		d.Set("severity", props.Severity)
+		d.Set("implementation_effort", props.ImplementationEffort)
+		d.Set("is_preview", props.Preview)
+		d.Set("remediation_description", props.RemediationDescription)
+		d.Set("user_impact", props.UserImpact)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go:229]: updated

File: azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+	}
+
+	d.Set("name", id.AssessmentMetadataName)
+
+	if props := resp.AssessmentMetadataProperties; props != nil {
+		d.Set("assessment_type", props.AssessmentType)
+		d.Set("description", props.Description)
+		d.Set("display_name", props.DisplayName)
+		d.Set("severity", props.Severity)
+		d.Set("implementation_effort", props.ImplementationEffort)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go:226]: updated

File: azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("retrieving Security Center Assessment Metadata %q : %+v", id.AssessmentMetadataName, err)
+	}
+
+	d.Set("name", id.AssessmentMetadataName)
+
+	if props := resp.AssessmentMetadataProperties; props != nil {
+		d.Set("assessment_type", props.AssessmentType)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go:222]: updated

File: azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+		d.Set("description", props.Description)
+		d.Set("display_name", props.DisplayName)
+		d.Set("severity", props.Severity)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go:225]: updated

File: azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+
+	name := d.Get("name").(string)
+
+	id := parse.NewAssessmentMetadataID(subscriptionId, name)
+
+	if d.IsNewResource() {
+		existing, err := client.GetInSubscription(ctx, name)
+		if err != nil {
+			if !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("checking for present of existing Security Center Assessment Metadata %q : %+v", name, err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go:148]: updated

File: azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+	if v, ok := d.GetOk("remediation_description"); ok {
+		params.AssessmentMetadataProperties.RemediationDescription = utils.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("user_impact"); ok {
+		params.AssessmentMetadataProperties.UserImpact = security.UserImpact(v.(string))
+	}
+
+	if _, err := client.CreateInSubscription(ctx, name, params); err != nil {
+		return fmt.Errorf("creating/updating Security Center Assessment Metadata %q : %+v", name, err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go:191]: updated

File: website/docs/r/security_center_assessment_metadata.html.markdown [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+
+* `implementation_effort` - (Optional) The implementation effort which is used to remediate the Security Center Assessment. Possible values are `Low`, `Moderate` and `High`.
+
+* `is_preview` - (Optional) Is the Security Center Assessment in preview release status?
+
+* `remediation_description` - (Optional) The description which is used to mitigate the security issue.
+
+* `threats` - (Optional) A list of the threat impacts for the Security Center Assessment. Possible values are `AccountBreach`, `DataExfiltration`, `DataSpillage`, `DenialOfService`, `ElevationOfPrivilege`, `MaliciousInsider`, `MissingCoverage` and `ThreatResistance`.
+
+* `user_impact` - (Optional) The user impact of the Security Center Assessment. Possible values are `Low`, `Moderate` and `High`.
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply website/docs/r/security_center_assessment_metadata.html.markdown:49]: Based on previous testing, there is no default value when this property isn't specified.

File: azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go [hunk 17]
```diff
@@ -0,0 +1,258 @@
...
+			return err
+		}),
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.IsUUID,
+			},
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go]: Updated.

File: azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+
+			// The `category` property doesn't take effect at the service side since the property name is incorrect and it should be `categories`.
+			// To implement this property once the bug is fixed.
+			// BUG: https://github.com/Azure/azure-rest-api-specs/issues/12297
+		},
+	}
+}
+
+func resourceArmSecurityCenterAssessmentMetadataCreateUpdate(d *schema.ResourceData, meta interface{}) error {
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/securitycenter/security_center_assessment_metadata_resource.go]: Updated

File: website/docs/r/security_center_assessment_metadata.html.markdown [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+
+* `display_name` - (Required) The user-friendly display name of the Security Center Assessment.
+
+* `severity` - (Required) The severity level of the Security Center Assessment. Possible values are `Low`, `Medium` and `High`.
+
+---
+
+* `implementation_effort` - (Optional) The implementation effort which is used to remediate the Security Center Assessment. Possible values are `Low`, `Moderate` and `High`.
+
+* `is_preview` - (Optional) Is the Security Center Assessment in preview release status?
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply website/docs/r/security_center_assessment_metadata.html.markdown:40]: After confirmed with service team, it's a read-only property.

**Comment by neil-yechenwei** [comment]
@tombuildsstuff , thanks for your comments. I've updated code. Please have an another look.

**Review by katbyte — APPROVED** [review]
Thanks @neil-yechenwei - this LGTM 👍

**Pull request merged**